### PR TITLE
expect: Add matcher `toBeInRange()` to validate if array elements are within the specified range

### PR DESCRIPTION
**Pull request by mayankshukla94**
### What

In this PR, I have created a new matcher `toBeInRange`. 

### Why

Recently, I came across a use case where I had to validate if `Year` dropdown has options within the specified range. When I checked for assertions, then no such assertion found where I can specify the range and validate or assert. 

### Housekeeping

- [x] Unit tests
- [x] Documentation is up to date
- [x] No additional lint warnings
- [x] [Typescript definitions](https://github.com/jest-community/jest-extended/blob/main/types/index.d.ts) are added/updated where relevant

### CODE REVIEW

**Comment by mayankshukla94** [comment]
> @keeganwitt @mattphillips thoughts?

@keeganwitt @mattphillips Could you please provide your input?

File: README.md [hunk 1]
```diff
@@ -390,6 +391,17 @@ test('passes when any value in array pass given predicate', () => {
 });
 ```
 
+#### .toBeWithinRange(min, max)
+
+Use `.toBeWithinRange` when you want to check if array contains elements within a given range.
```
Comment (keeganwitt) [inline README.md:396]: We should probably document here that the numbers are both inclusive.

**Comment by keeganwitt** [comment]
This seems reasonable to me. We could make the inclusive vs exclusive more flexible, but I don't think that'd add much value, and being inclusive is what I'd kinda expect.

File: README.md (hunk 1 above)
Comment (mayankshukla94) [reply README.md]: @keeganwitt I have updated it.
Thanks!

File: test/matchers/toBeWithinRange.test.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import * as matcher from 'src/matchers/toBeWithinRange';
+
+expect.extend(matcher);
+
+describe('.toBeWithinRange', () => {
+  test('passes when given array is within range', () => {
+    expect([4, 5, 7, 9]).toBeWithinRange(4, 9);
```
Comment (keeganwitt) [inline test/matchers/toBeWithinRange.test.js:7]: I don't think an array should be passed here.

File: test/matchers/toBeWithinRange.test.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  });
+
+  test('fails when given array is not within a given range', () => {
+    expect(() => expect([4, 5, 7, 9]).toBeWithinRange(4, 8)).toThrowErrorMatchingSnapshot();
+  });
+});
+
+describe('.not.toBeWithinRange', () => {
+  test('passes when given array is not within the given range', () => {
+    expect([12, 13, 15, 17]).not.toBeWithinRange(4, 9);
```
Comment (keeganwitt) [inline test/matchers/toBeWithinRange.test.js:17]: Same here.

File: types/index.d.ts [hunk 4]
```diff
@@ -633,6 +633,14 @@ declare namespace jest {
      */
     toBeWithin(start: number, end: number): any;
 
+    /**
+     * Use `.toBeWithinRange` when checking if an array has elements in range min (inclusive) and max (inclusive).
+     *
+     * @param min
+     * @param max
+     */
+    toBeWithinRange(min: number, max: number): any;
```
Comment (keeganwitt) [inline types/index.d.ts:642]: Please rebase to replace the `any`. Also add a signature to the `Matchers` interface.

File: test/matchers/toBeWithinRange.test.js (hunk 2 above)
Comment (mayankshukla94) [reply test/matchers/toBeWithinRange.test.js:7]: Do you mean

```
const arrayList = [4, 5, 7, 9];
expect(arrayList).toBeWithinRange(4, 9);
```

File: types/index.d.ts (hunk 4 above)
Comment (mayankshukla94) [reply types/index.d.ts]: Done

File: test/matchers/toBeWithinRange.test.js (hunk 2 above)
Comment (keeganwitt) [reply test/matchers/toBeWithinRange.test.js:7]: No, I thought the type of expected was a number, not an array, but now I see you have `actual.find()`.

I'm wondering if this should accept a number instead (or maybe in addition).

Comment (mayankshukla94) [reply test/matchers/toBeWithinRange.test.js:7]: > I'm wondering if this should accept a number instead

There's already an assertion for that called as `toBeWithin()`. So do we really need this `toBeWithinRange()` as well to accept a number?

**Comment by mayankshukla94** [comment]
@mattphillips Thank you for the detailed review. I have made the changes. Please review again.

File: src/matchers/toBeInRange.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
+export function toBeInRange(actual, min, max) {
+  const { printReceived, printExpected, matcherHint } = this.utils;
+
+  const element = actual.find(option => option < min || option >= max);
```
Comment (keeganwitt) [inline src/matchers/toBeInRange.js:4]: Should there also be a check if the element isn't a number?

**Comment by mayankshukla94** [comment]
Hello @keeganwitt @mattphillips,

Could you please review this again?

Thanks!

**Comment by mayankshukla94** [comment]
@mattphillips @keeganwitt I have resolved the conflicts. Please have a look again.

**Comment by mayankshukla94** [comment]
Hello @SimenB @keeganwitt @mattphillips 

Any update on this?

**Comment by keeganwitt** [comment]
Sorry, I haven't had a chance to get back to this. I'll take a look this week.

**Comment by keeganwitt** [comment]
A few more things are needed before this is ready:
1. https://github.com/jest-community/jest-extended/blob/main/website/docs/matchers/index.md should be updated to include this new matcher.
1. There are two places in [index.d.ts](https://github.com/jest-community/jest-extended/blob/main/types/index.d.ts) that get updated every time a new matcher is added: the `Matchers` interface and `Expect` interface. This PR only has the latter.

**Comment by mayankshukla94** [comment]
> A few more things are needed before this is ready:
> 1. https://github.com/jest-community/jest-extended/blob/main/website/docs/matchers/index.md should be updated to include this new matcher.
> 1. There are two places in [index.d.ts](https://github.com/jest-community/jest-extended/blob/main/types/index.d.ts) that get updated every time a new matcher is added: the `Matchers` interface and `Expect` interface. This PR only has the latter.

Sure will update.

**Comment by keeganwitt** [comment]
> > A few more things are needed before this is ready:
> > 
> > 1. https://github.com/jest-community/jest-extended/blob/main/website/docs/matchers/index.md should be updated to include this new matcher.
> > 2. There are two places in [index.d.ts](https://github.com/jest-community/jest-extended/blob/main/types/index.d.ts) that get updated every time a new matcher is added: the `Matchers` interface and `Expect` interface. This PR only has the latter.
> 
> Sure will update.

Sounds good! Thanks for your persistence!

**Comment by mayankshukla94** [comment]
@keeganwitt I have updated now. Please review

File: website/docs/matchers/Array.mdx [hunk 6]
```diff
@@ -105,3 +105,14 @@ Use `.toSatisfyAny` when you want to use a custom matcher by supplying a predica
   expect([2, 4, 8, 12]).not.toSatisfyAny(isOdd);
 });`}
 </TestFile>
+
+### .toBeInRange(min, max)
+
+Use `.toBeInRange` when you want to check if the given array contains numbers within the specified range.
```
Comment (keeganwitt) [inline website/docs/matchers/Array.mdx:111]: I'd include that min is exclusive, and max is inclusive here.

Comment (mayankshukla94) [reply website/docs/matchers/Array.mdx:111]: Not sure if @mattphillips has been notified because this comment is edited.

Comment (keeganwitt) [reply website/docs/matchers/Array.mdx:111]: An example of how to word the inclusive/exclusive behavior is the toBeWithin documentation
> Use `.toBeWithin` when checking if a number is in between the given bounds of: start (inclusive) and end (exclusive).

Comment (mayankshukla94) [reply website/docs/matchers/Array.mdx]: @keeganwitt I have updated the description now.

**Review by keeganwitt — APPROVED** [review]

File: website/docs/matchers/Array.mdx (hunk 6 above)
Comment (mayankshukla94) [reply website/docs/matchers/Array.mdx]: @keeganwitt I'm not able to request a review from @mattphillips. Could you please help here?

Comment (keeganwitt) [reply website/docs/matchers/Array.mdx]: I was going to wait another day-ish, and if he doesn't respond, I'll just merge it.

Comment (mayankshukla94) [reply website/docs/matchers/Array.mdx]: @keeganwitt Could you please merge this PR to main branch if all looks good to you?

Comment (keeganwitt) [reply website/docs/matchers/Array.mdx]: Yea. We can always rename before cutting the next release if there's a better name that we should have used.

**Pull request merged**